src/java/org/apache/cassandra/db/filter/ColumnFilter.java

```diff
@@ -505,9 +504,16 @@ public ColumnFilter deserialize(DataInputPlus in, int version, CFMetaData metada
                 }
             }
 
-            // See CASSANDRA-15833
-            if (version <= MessagingService.VERSION_3014 && isFetchAll)
+            // Since nodes with and without CASSANDRA-10657 are not distinguishable by messaging version, we need to
+            // check whether there are any nodes running pre CASSANDRA-10657 Cassandra and apply conversion to the
+            // column filter so that it is interpreted in the same way as on the nodes without CASSANDRA-10657.
+            //
+            // see CASSANDRA-10657, CASSANDRA-15833, CASSANDRA-16415
+            if (isFetchAll && queried != null && Gossiper.instance.isAnyNodeOn30())
```

Comment:
Thanks for the comment. 

The previous condition was wrong. Cassandra 3.11 uses `MessagingService.VERSION_3014`. Checking `version <= MessagingService.VERSION_3014` incorrectly nullifies `queried`. 

The new condition looks good. 

However, my concern is that gossip is not very reliable. (Not saying the fix is bad. It is better than the previous one) Checking `Gossiper.instance.isAnyNodeOn30()` might still leave a small time window of digest mismatch during upgrade. The mixed mode scenario happens during upgrading a cluster. We can assume the value of `has30Node` on each instance only changes in 1 direction, i.e. true --> false. It will never be the other way around. However, at a given time point, the status of the ring can be inconsistent among the nodes due to the gossip mechanism. Some nodes can learn that all nodes are now in 3.11 (which is the fact), but the others disagree due to the delay. In such scenario, if the coordinator and replica 1 (R1) thinks all nodes are already in 3.11, but replica 2 (R2) thinks that there are nodes still in 3.0, the digest returned from R1 mismatches the one from R2.  

I think a small time window is tolerable. Read latencies can go up. But eventually, when the gossip status is propagated to the entire cluster, the mismatch should disappear. WDYT?

Reply:
`anyNodeOn30` is not a flag which changes one way. It is initially set to `false`, which means that in case of the upgrade it will change to `true` and then eventually back to `false`. This is implemented differently on `trunk` but I didn't want to change the behaviour of this flag here as more things depend on it and it is just the patch release. There will be obviously small time gap when we still get digest mismatches, but it is still better than having them during whole upgrade. If the flag was set initially to `true`, it would be better - initially the cluster would assume 3.0 behaviour.

For the future releases, I think we could make messaging version correspond to version family (which means incrementing it with each new minor release) to ensure we wouldn't be forced to base on Gossiper.

Please have a look on trunk PR because it had to be more complicated as we need to distinguish whether we upgrade from pre 3.4 or 3.4~4.0